src/lib/tuning/greedy_selector.cpp

```diff
@@ -0,0 +1,143 @@
+    cost_delta += (*choice)->reject_cost();
+    ++choice;
+  }
+
+  if (cost_delta <= required_cost_delta && desirability_delta >= acceptible_desirability_delta) {
+    return choice;
+  }
+  return sorted_choices.end();
+}
+
+std::vector<std::shared_ptr<TuningOperation>> GreedySelector::select(
+    const std::vector<std::shared_ptr<TuningChoice>>& choices, float cost_budget) {
+  std::vector<std::shared_ptr<TuningOperation>> operations;
+  operations.reserve(choices.size());
+  // Assumption: cost() >= 0 ==> accept_cost() >= 0 && current_cost() >= 0 && reject_cost() <= 0
+
+  // Accumulate absolute cost balance from currently chosen choices.
+  float cost_balance = 0.0f;
+  for (const auto& choice : choices) {
+    cost_balance += choice->current_cost();
+  }
+
+  // Desirability balance is always relative to current system state.
+  float desirability_balance = 0.0f;
+
+  // Sort choices:
+  //  1) by reject_cost() ascending
+  //  2) by accept_desirability() ascending stable
+  // The most desirable choice is at back() and among equaly desirable choices
+  // the least costly choices to accept are nearer back().
+  std::list<std::shared_ptr<TuningChoice>> sorted_choices(choices.size());
+  std::copy(choices.cbegin(), choices.cend(), sorted_choices.begin());
+  sorted_choices.sort([](std::shared_ptr<TuningChoice> lhs, std::shared_ptr<TuningChoice> rhs) {
+    return lhs->reject_cost() < rhs->reject_cost();
+  });
+  sorted_choices.sort([](std::shared_ptr<TuningChoice> lhs, std::shared_ptr<TuningChoice> rhs) {
+    return lhs->accept_desirability() < rhs->accept_desirability();
+  });
+
+  // If current state exceeds cost_budget,
+  // reject the least desirable choices to reduce cost_balance.
+  if (cost_balance > cost_budget) {
+    LOG_INFO("Cost balance of " << cost_balance << " exceeds budget of " << cost_budget);
+    auto sacrifice_until = sacrifice_choices(sorted_choices, cost_budget - cost_balance);
+    if (sacrifice_until == sorted_choices.end()) {
+      LOG_WARN("Cost budget is impossible to maintain. No Operations are performed.");
+      return operations;
+    }
+    for (auto choice = sorted_choices.begin(); choice != sacrifice_until; ++choice) {
+      LOG_DEBUG(" ! Reject " << **choice << " to reduce cost balance");
+      operations.push_back((*choice)->reject());
+      cost_balance += (*choice)->reject_cost();
+      desirability_balance += (*choice)->reject_desirability();
+    }
+    sorted_choices.erase(sorted_choices.begin(), sacrifice_until);
+  }
+
+  // Select the most desirable operations first
+  // always maintaining the cost budget
+  while (sorted_choices.size() > 0) {
+    if (sorted_choices.front()->reject_desirability() > sorted_choices.back()->accept_desirability()) {
+      LOG_DEBUG("Rejecting " << *sorted_choices.front() << " is most beneficial.");
+      // Rejecting a choice can only reduce cost_balance and never exceed cost_budget (as cost()>=0)
```

Comment:
Could you clarify this comment?

Reply:
I hope this is better now:
> Rejecting a choice can only reduce cost_balance (i.e. free up resources) and never exceed cost_budget (no new costs will be added) as cost() >= 0: Rejecting a choice that already exists yields ("gives back") the cost that was paid for it. Rejecting a choice that does not yet exist cannot cost anything since nothing will be changed.